<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="PACS">
<title>Vignette 1. Cell type annotation with PACS â€¢ PACS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Vignette 1. Cell type annotation with PACS">
<meta property="og:description" content="PACS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PACS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Vignette_1_PACS_cell_type_annotation.html">Vignette 1. Cell type annotation with PACS</a>
    <a class="dropdown-item" href="../articles/Vignette_2_PACS_differential_test.html">Vignette 2. Differentially Accessible Region (DAR) Identification with PACS</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Vignette 1. Cell type annotation with PACS</h1>
            
            <h4 data-toc-skip class="date">2023-08-21</h4>
      
      
      <div class="d-none name"><code>Vignette_1_PACS_cell_type_annotation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="content">Content<a class="anchor" aria-label="anchor" href="#content"></a>
</h2>
<p>This Vignette contains an example to conduct cell type annotation
with PACS. To conduct cell type annotation, we need to have a reference
dataset with cell type annotated, and a new dataset where we want to
infer the cell type labels for each cell.</p>
</div>
<div class="section level2">
<h2 id="required-libraries">Required libraries<a class="anchor" aria-label="anchor" href="#required-libraries"></a>
</h2>
<p>Please make sure the following libraries are installed and loaded for
the analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://r-datatable.com" class="external-link">"data.table"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://Matrix.R-forge.R-project.org" class="external-link">"Matrix"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://zhen-miao.github.io/PICsnATAC/" class="external-link">"PICsnATAC"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://zhen-miao.github.io/PACS/">"PACS"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/RfastOfficial/Rfast" class="external-link">"Rfast"</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="input-files">Input files<a class="anchor" aria-label="anchor" href="#input-files"></a>
</h2>
<p>The PACS cell type annotation framework requires the following input
files:</p>
<ul>
<li>Reference peak by cell matrix (<code>ref_mat</code>)</li>
<li>Cell type labels for the reference data
(<code>cell_types</code>)</li>
<li>Target peak by cell matrix that we want to obtain cell type labels
(<code>tar_mat</code>)</li>
</ul>
<p>Note, the two data matrices should have the same set of features
(peaks), and please refer to our <a href="https://github.com/Zhen-Miao/PIC-snATAC" class="external-link"><code>PIC_snATAC</code>
package</a> for obtaining data matrix. Clustering and annotation can be
conducted with Seurat, ArchR, snapATAC, or other pipelines.</p>
</div>
<div class="section level2">
<h2 id="cell-specific-capturing-rate-estimation">Cell-specific capturing rate estimation<a class="anchor" aria-label="anchor" href="#cell-specific-capturing-rate-estimation"></a>
</h2>
<p>As described in the manuscript, in the snATAC-seq data, each cell may
have largely variant capturing rate (sequencing depth), and here, we
will compute the caputuring rate in each cell while separating the
effect of cell type-specific open probability</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cell_types</span><span class="op">)</span></span>
<span><span class="va">r_by_ct_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://zhen-miao.github.io/PICsnATAC/reference/get_r_by_ct_mat_pq.html" class="external-link">get_r_by_ct_mat_pq</a></span><span class="op">(</span></span>
<span>  cell_type_set <span class="op">=</span> <span class="va">ctypes</span>,</span>
<span>  r_by_c <span class="op">=</span> <span class="va">ref_mat</span>,</span>
<span>  cell_type_labels <span class="op">=</span> <span class="va">cell_types</span>,</span>
<span>  n_features_per_cell <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">ref_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The object <code>r_by_ct_out</code> is a list with two elements,</p>
<ul>
<li>
<strong>p_by_t</strong> Peak by cell type matrix, each element
represents the open probability of the peak in the corresponding cell
type</li>
<li>
<strong>q_vec</strong> A vector of cell-specific capturing rate</li>
</ul>
<p>The first element will be the input for annotating cell type labels
of new dataset</p>
</div>
<div class="section level2">
<h2 id="identify-informative-features">Identify informative features<a class="anchor" aria-label="anchor" href="#identify-informative-features"></a>
</h2>
<p>We only choose a subset of features that are informative for
distinguish different cell types. A rule of thumb is to choose features
based on their standard deviation across cell types.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_sd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu">Rfast</span><span class="fu">::</span><span class="fu">rowVars</span><span class="op">(</span><span class="va">r_by_ct_out</span><span class="op">$</span><span class="va">p_by_t</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pk_sel</span> <span class="op">&lt;-</span> <span class="va">p_sd</span> <span class="op">&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile</a></span><span class="op">(</span><span class="va">p_sd</span>, <span class="fl">0.5</span><span class="op">)</span> <span class="co">## top 50% highly variable features</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cell-type-annotation">Cell type annotation<a class="anchor" aria-label="anchor" href="#cell-type-annotation"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## get the likelihood for each cell type</span></span>
<span><span class="va">est_mat</span> <span class="op">&lt;-</span> <span class="fu">estimate_label_no_cap_rate</span><span class="op">(</span></span>
<span>  r_by_t <span class="op">=</span> <span class="va">r_by_ct_out</span><span class="op">$</span><span class="va">p_by_t</span><span class="op">[</span><span class="va">pk_sel</span>, <span class="op">]</span>,</span>
<span>  in_r_by_c <span class="op">=</span> <span class="va">tar_mat</span><span class="op">[</span><span class="va">pk_sel</span>, <span class="op">]</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## identify the most probable cell type</span></span>
<span><span class="va">esti_ctype_fir</span> <span class="op">&lt;-</span> <span class="fu">Rfast</span><span class="fu">::</span><span class="fu">rownth</span><span class="op">(</span><span class="va">est_mat</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, length <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">est_mat</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  num.of.nths <span class="op">=</span> <span class="fl">1</span>, descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  index.return <span class="op">=</span> <span class="cn">TRUE</span>, parallel <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">esti_ctype_labels</span> <span class="op">&lt;-</span> <span class="va">ctypes</span><span class="op">[</span><span class="va">esti_ctype_fir</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>If you used PACS in your analysis, please cite our manuscript:</p>
<p>Miao, Z., Wang, J., Park, K., Kuang, D., &amp; Kim, J. (2023).
Model-based compound hypothesis testing for snATAC-seq data with PACS.
<em>bioRxiv</em>, 2023-07.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Zhen Miao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
